internal/transport/http2_client.go

```diff
@@ -702,7 +709,14 @@ func (e NewStreamError) Error() string {
 // streams.  All non-nil errors returned will be *NewStreamError.
 func (t *http2Client) NewStream(ctx context.Context, callHdr *CallHdr) (*Stream, error) {
 	ctx = peer.NewContext(ctx, t.getPeer())
-	headerFields, err := t.createHeaderFields(ctx, callHdr)
+
+	dupCallHdr := *callHdr
+	// replace host with the actual server name, if it is exist and unmatch
```

Comment:
Nit: Please treat comment sentences like regular sentences.

Something like this might be more useful here:
```golang
// ServerName field of the resolver returned address takes precedence over
// Host field of CallHdr to determine the :authority header. This is because,
// the ServerName field takes precedence for server authentication during
// TLS handshake, and the :authority header should match the value used
// for server authentication.
```